File: app/src/main/java/dev/ragnarok/fenrir/model/PhotoSizes.java
package dev.ragnarok.fenrir.model;

import static dev.ragnarok.fenrir.util.Utils.firstNonNull;

import android.os.Parcel;
import android.os.Parcelable;

import androidx.annotation.Keep;

import dev.ragnarok.fenrir.module.parcel.ParcelNative;


@Keep
public class PhotoSizes implements Parcelable, ParcelNative.ParcelableNative {

    public static final Creator<PhotoSizes> CREATOR = new Creator<PhotoSizes>() {
        @Override
        public PhotoSizes createFromParcel(Parcel in) {
            return new PhotoSizes(in);
        }

        @Override
        public PhotoSizes[] newArray(int size) {
            return new PhotoSizes[size];
        }
    };
    public static final ParcelNative.Creator<PhotoSizes> NativeCreator = PhotoSizes::new;
    private Size s;
    private Size m;
    private Size x;
    private Size o;
    private Size p;
    private Size q;
    private Size r;
    private Size y;
    private Size z;
    private Size w;

    public PhotoSizes() {

    }

    protected PhotoSizes(ParcelNative in) {
        s = in.readParcelable(Size.NativeCreator);
        m = in.readParcelable(Size.NativeCreator);
        x = in.readParcelable(Size.NativeCreator);
        o = in.readParcelable(Size.NativeCreator);
        p = in.readParcelable(Size.NativeCreator);
        q = in.readParcelable(Size.NativeCreator);
        r = in.readParcelable(Size.NativeCreator);
        y = in.readParcelable(Size.NativeCreator);
        z = in.readParcelable(Size.NativeCreator);
        w = in.readParcelable(Size.NativeCreator);
    }

    protected PhotoSizes(Parcel in) {
        s = in.readParcelable(Size.class.getClassLoader());
        m = in.readParcelable(Size.class.getClassLoader());
        x = in.readParcelable(Size.class.getClassLoader());
        o = in.readParcelable(Size.class.getClassLoader());
        p = in.readParcelable(Size.class.getClassLoader());
        q = in.readParcelable(Size.class.getClassLoader());
        r = in.readParcelable(Size.class.getClassLoader());
        y = in.readParcelable(Size.class.getClassLoader());
        z = in.readParcelable(Size.class.getClassLoader());
        w = in.readParcelable(Size.class.getClassLoader());
    }

    public static PhotoSizes empty() {
        return new PhotoSizes();
    }

    public Size getS() {
        return s;
    }

    public PhotoSizes setS(Size s) {
        this.s = s;
        return this;
    }

    public Size getM() {
        return m;
    }

    public PhotoSizes setM(Size m) {
        this.m = m;
        return this;
    }

    public Size getX() {
        return x;
    }

    public PhotoSizes setX(Size x) {
        this.x = x;
        return this;
    }

    public Size getO() {
        return o;
    }

    public PhotoSizes setO(Size o) {
        this.o = o;
        return this;
    }

    public Size getP() {
        return p;
    }

    public PhotoSizes setP(Size p) {
        this.p = p;
        return this;
    }

    public Size getQ() {
        return q;
    }

    public PhotoSizes setQ(Size q) {
        this.q = q;
        return this;
    }

    public Size getR() {
        return r;
    }

    public PhotoSizes setR(Size r) {
        this.r = r;
        return this;
    }

    public Size getY() {
        return y;
    }

    public PhotoSizes setY(Size y) {
        this.y = y;
        return this;
    }

    public Size getZ() {
        return z;
    }

    public PhotoSizes setZ(Size z) {
        this.z = z;
        return this;
    }

    public Size getW() {
        return w;
    }

    public PhotoSizes setW(Size w) {
        this.w = w;
        return this;
    }

    public Size getMaxSize(boolean excludeNonAspectRatio) {
        return excludeNonAspectRatio ? firstNonNull(w, z, y, x, m, s) : firstNonNull(w, z, y, r, q, p, o, x, m, s);
    }

    public Size getSize(@PhotoSize int max, boolean excludeNonAspectRatio) {
        switch (max) {
            case PhotoSize.S:
                return s;
            case PhotoSize.M:
                return firstNonNull(m, s);
            case PhotoSize.X:
                return firstNonNull(x, m, s);
            case PhotoSize.O:
                return excludeNonAspectRatio ? firstNonNull(x, m, s)
                        : firstNonNull(o, x, m, s);
            case PhotoSize.P:
                return excludeNonAspectRatio ? firstNonNull(x, m, s)
                        : firstNonNull(p, o, x, m, s);
            case PhotoSize.Q:
                return excludeNonAspectRatio ? firstNonNull(x, m, s)
                        : firstNonNull(q, p, o, x, m, s);
            case PhotoSize.R:
                return excludeNonAspectRatio ? firstNonNull(x, m, s)
                        : firstNonNull(r, q, p, o, x, m, s);
            case PhotoSize.Y:
                return excludeNonAspectRatio ? firstNonNull(y, x, m, s)
                        : firstNonNull(y, r, q, p, o, x, m, s);
            case PhotoSize.Z:
                return excludeNonAspectRatio ? firstNonNull(z, y, x, m, s)
                        : firstNonNull(z, y, r, q, p, o, x, m, s);
            case PhotoSize.W:
                return excludeNonAspectRatio ? firstNonNull(w, z, y, x, m, s)
                        : firstNonNull(w, z, y, r, q, p, o, x, m, s);
            default:
                throw new IllegalArgumentException("Invalid max photo size: " + max);
        }
    }

    public String getUrlForSize(@PhotoSize int maxSize, boolean excludeNonAspectRatio) {
        Size s = getSize(maxSize, excludeNonAspectRatio);
        return s == null ? null : s.url;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel parcel, int i) {
        parcel.writeParcelable(s, i);
        parcel.writeParcelable(m, i);
        parcel.writeParcelable(x, i);
        parcel.writeParcelable(o, i);
        parcel.writeParcelable(p, i);
        parcel.writeParcelable(q, i);
        parcel.writeParcelable(r, i);
        parcel.writeParcelable(y, i);
        parcel.writeParcelable(z, i);
        parcel.writeParcelable(w, i);
    }

    public boolean isEmpty() {
        return firstNonNull(s, m, x, o, p, q, r, y, z, w) == null;
    }

    @Override
    public void writeToParcelNative(ParcelNative parcel) {
        parcel.writeParcelable(s);
        parcel.writeParcelable(m);
        parcel.writeParcelable(x);
        parcel.writeParcelable(o);
        parcel.writeParcelable(p);
        parcel.writeParcelable(q);
        parcel.writeParcelable(r);
        parcel.writeParcelable(y);
        parcel.writeParcelable(z);
        parcel.writeParcelable(w);
    }

    @Keep
    public static final class Size implements Parcelable, ParcelNative.ParcelableNative {

        public static final Creator<Size> CREATOR = new Creator<Size>() {
            @Override
            public Size createFromParcel(Parcel in) {
                return new Size(in);
            }

            @Override
            public Size[] newArray(int size) {
                return new Size[size];
            }
        };
        public static final ParcelNative.Creator<Size> NativeCreator = Size::new;
        private int width;
        private int height;
        private String url;

        @SuppressWarnings("unused")
        public Size() {
        }

        public Size(int w, int h, String url) {
            width = w;
            height = h;
            this.url = url;
        }

        Size(ParcelNative in) {
            width = in.readInt();
            height = in.readInt();
            url = in.readString();
        }

        Size(Parcel in) {
            width = in.readInt();
            height = in.readInt();
            url = in.readString();
        }

        public String getUrl() {
            return url;
        }

        public int getW() {
            return width;
        }

        public int getH() {
            return height;
        }

        @Override
        public int describeContents() {
            return 0;
        }

        @Override
        public void writeToParcel(Parcel dest, int flags) {
            dest.writeInt(width);
            dest.writeInt(height);
            dest.writeString(url);
        }

        @Override
        public void writeToParcelNative(ParcelNative dest) {
            dest.writeInt(width);
            dest.writeInt(height);
            dest.writeString(url);
        }
    }
}

File: app_fenrir/src/main/kotlin/dev/ragnarok/fenrir/model/PhotoSizes.kt
package dev.ragnarok.fenrir.model

import android.os.Parcel
import android.os.Parcelable
import androidx.annotation.Keep
import dev.ragnarok.fenrir.module.parcel.ParcelNative
import dev.ragnarok.fenrir.readTypedObjectCompat
import dev.ragnarok.fenrir.util.Utils.firstNonNull
import dev.ragnarok.fenrir.writeTypedObjectCompat
import kotlinx.serialization.Serializable

@Keep
@Serializable
class PhotoSizes : Parcelable, ParcelNative.ParcelableNative {
    private var s: Size? = null
    private var m: Size? = null
    private var x: Size? = null
    private var o: Size? = null
    private var p: Size? = null
    private var q: Size? = null
    private var r: Size? = null
    private var k: Size? = null
    private var l: Size? = null
    private var y: Size? = null
    private var z: Size? = null
    private var w: Size? = null

    constructor()
    internal constructor(parcel: ParcelNative) {
        s = parcel.readParcelable(Size.NativeCreator)
        m = parcel.readParcelable(Size.NativeCreator)
        x = parcel.readParcelable(Size.NativeCreator)
        o = parcel.readParcelable(Size.NativeCreator)
        p = parcel.readParcelable(Size.NativeCreator)
        q = parcel.readParcelable(Size.NativeCreator)
        r = parcel.readParcelable(Size.NativeCreator)
        y = parcel.readParcelable(Size.NativeCreator)
        z = parcel.readParcelable(Size.NativeCreator)
        w = parcel.readParcelable(Size.NativeCreator)
        k = parcel.readParcelable(Size.NativeCreator)
        l = parcel.readParcelable(Size.NativeCreator)
    }

    internal constructor(parcel: Parcel) {
        s = parcel.readTypedObjectCompat(Size.CREATOR)
        m = parcel.readTypedObjectCompat(Size.CREATOR)
        x = parcel.readTypedObjectCompat(Size.CREATOR)
        o = parcel.readTypedObjectCompat(Size.CREATOR)
        p = parcel.readTypedObjectCompat(Size.CREATOR)
        q = parcel.readTypedObjectCompat(Size.CREATOR)
        r = parcel.readTypedObjectCompat(Size.CREATOR)
        y = parcel.readTypedObjectCompat(Size.CREATOR)
        z = parcel.readTypedObjectCompat(Size.CREATOR)
        w = parcel.readTypedObjectCompat(Size.CREATOR)
        k = parcel.readTypedObjectCompat(Size.CREATOR)
        l = parcel.readTypedObjectCompat(Size.CREATOR)
    }

    fun getS(): Size? {
        return s
    }

    fun setS(s: Size?): PhotoSizes {
        this.s = s
        return this
    }

    fun getM(): Size? {
        return m
    }

    fun setM(m: Size?): PhotoSizes {
        this.m = m
        return this
    }

    fun getX(): Size? {
        return x
    }

    fun setX(x: Size?): PhotoSizes {
        this.x = x
        return this
    }

    fun getO(): Size? {
        return o
    }

    fun setO(o: Size?): PhotoSizes {
        this.o = o
        return this
    }

    fun getP(): Size? {
        return p
    }

    fun setP(p: Size?): PhotoSizes {
        this.p = p
        return this
    }

    fun getQ(): Size? {
        return q
    }

    fun setQ(q: Size?): PhotoSizes {
        this.q = q
        return this
    }

    fun getR(): Size? {
        return r
    }

    fun setR(r: Size?): PhotoSizes {
        this.r = r
        return this
    }

    fun getY(): Size? {
        return y
    }

    fun setY(y: Size?): PhotoSizes {
        this.y = y
        return this
    }

    fun getZ(): Size? {
        return z
    }

    fun setZ(z: Size?): PhotoSizes {
        this.z = z
        return this
    }

    fun getW(): Size? {
        return w
    }

    fun setW(w: Size?): PhotoSizes {
        this.w = w
        return this
    }

    fun getK(): Size? {
        return k
    }

    fun setK(k: Size?): PhotoSizes {
        this.k = k
        return this
    }

    fun getL(): Size? {
        return l
    }

    fun setL(l: Size?): PhotoSizes {
        this.l = l
        return this
    }

    fun getMaxSize(excludeNonAspectRatio: Boolean): Size? {
        return if (excludeNonAspectRatio) firstNonNull(
            w,
            z,
            y,
            x,
            m,
            s
        ) else firstNonNull(w, z, y, r, q, p, o, x, m, s)
    }

    fun getSize(@PhotoSize max: Int, excludeNonAspectRatio: Boolean): Size? {
        return when (max) {
            PhotoSize.S -> s
            PhotoSize.M -> firstNonNull(m, s)
            PhotoSize.X -> firstNonNull(x, m, s)
            PhotoSize.O -> if (excludeNonAspectRatio) firstNonNull(
                x,
                m,
                s
            ) else firstNonNull(o, x, m, s)

            PhotoSize.P -> if (excludeNonAspectRatio) firstNonNull(
                x,
                m,
                s
            ) else firstNonNull(p, o, x, m, s)

            PhotoSize.Q -> if (excludeNonAspectRatio) firstNonNull(
                x,
                m,
                s
            ) else firstNonNull(q, p, o, x, m, s)

            PhotoSize.R -> if (excludeNonAspectRatio) firstNonNull(
                x,
                m,
                s
            ) else firstNonNull(r, q, p, o, x, m, s)

            PhotoSize.Y -> if (excludeNonAspectRatio) firstNonNull(
                y,
                x,
                m,
                s
            ) else firstNonNull(y, r, q, p, o, x, m, s)

            PhotoSize.Z -> if (excludeNonAspectRatio) firstNonNull(
                z,
                y,
                x,
                m,
                s
            ) else firstNonNull(z, y, r, q, p, o, x, m, s)

            PhotoSize.W -> if (excludeNonAspectRatio) firstNonNull(
                w,
                z,
                y,
                x,
                m,
                s
            ) else firstNonNull(w, z, y, r, q, p, o, x, m, s)

            PhotoSize.K -> if (excludeNonAspectRatio) firstNonNull(
                k,
                y,
                x,
                m,
                s
            ) else firstNonNull(k, y, r, q, p, o, x, m, s)

            PhotoSize.L -> if (excludeNonAspectRatio) firstNonNull(
                l,
                k,
                y,
                x,
                m,
                s
            ) else firstNonNull(l, k, y, r, q, p, o, x, m, s)

            else -> throw IllegalArgumentException("Invalid max photo size: $max")
        }
    }

    fun getUrlForSize(@PhotoSize maxSize: Int, excludeNonAspectRatio: Boolean): String? {
        val s = getSize(maxSize, excludeNonAspectRatio)
        return s?.url
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(parcel: Parcel, i: Int) {
        parcel.writeTypedObjectCompat(s, i)
        parcel.writeTypedObjectCompat(m, i)
        parcel.writeTypedObjectCompat(x, i)
        parcel.writeTypedObjectCompat(o, i)
        parcel.writeTypedObjectCompat(p, i)
        parcel.writeTypedObjectCompat(q, i)
        parcel.writeTypedObjectCompat(r, i)
        parcel.writeTypedObjectCompat(y, i)
        parcel.writeTypedObjectCompat(z, i)
        parcel.writeTypedObjectCompat(w, i)
        parcel.writeTypedObjectCompat(k, i)
        parcel.writeTypedObjectCompat(l, i)
    }

    fun isEmpty(): Boolean {
        return firstNonNull(s, m, x, o, p, q, r, y, z, w, k, l) == null
    }

    override fun writeToParcelNative(dest: ParcelNative) {
        dest.writeParcelable(s)
        dest.writeParcelable(m)
        dest.writeParcelable(x)
        dest.writeParcelable(o)
        dest.writeParcelable(p)
        dest.writeParcelable(q)
        dest.writeParcelable(r)
        dest.writeParcelable(y)
        dest.writeParcelable(z)
        dest.writeParcelable(w)
        dest.writeParcelable(k)
        dest.writeParcelable(l)
    }

    @Keep
    @Serializable
    class Size : Parcelable, ParcelNative.ParcelableNative {
        private var width = 0
        private var height = 0
        var url: String? = null

        constructor()
        constructor(w: Int, h: Int, url: String?) {
            width = w
            height = h
            this.url = url
        }

        internal constructor(parcel: ParcelNative) {
            width = parcel.readInt()
            height = parcel.readInt()
            url = parcel.readString()
        }

        internal constructor(parcel: Parcel) {
            width = parcel.readInt()
            height = parcel.readInt()
            url = parcel.readString()
        }

        fun getW(): Int {
            return width
        }

        fun getH(): Int {
            return height
        }

        override fun describeContents(): Int {
            return 0
        }

        override fun writeToParcel(dest: Parcel, flags: Int) {
            dest.writeInt(width)
            dest.writeInt(height)
            dest.writeString(url)
        }

        override fun writeToParcelNative(dest: ParcelNative) {
            dest.writeInt(width)
            dest.writeInt(height)
            dest.writeString(url)
        }

        companion object {
            @JvmField
            val CREATOR: Parcelable.Creator<Size> = object : Parcelable.Creator<Size> {
                override fun createFromParcel(parcel: Parcel): Size {
                    return Size(parcel)
                }

                override fun newArray(size: Int): Array<Size?> {
                    return arrayOfNulls(size)
                }
            }
            val NativeCreator: ParcelNative.Creator<Size> =
                object : ParcelNative.Creator<Size> {
                    override fun readFromParcelNative(dest: ParcelNative): Size {
                        return Size(dest)
                    }

                }
        }
    }

    companion object {
        @JvmField
        val CREATOR: Parcelable.Creator<PhotoSizes> = object : Parcelable.Creator<PhotoSizes> {
            override fun createFromParcel(parcel: Parcel): PhotoSizes {
                return PhotoSizes(parcel)
            }

            override fun newArray(size: Int): Array<PhotoSizes?> {
                return arrayOfNulls(size)
            }
        }
        val NativeCreator: ParcelNative.Creator<PhotoSizes> =
            object : ParcelNative.Creator<PhotoSizes> {
                override fun readFromParcelNative(dest: ParcelNative): PhotoSizes {
                    return PhotoSizes(dest)
                }

            }
    }
}